pjmedia/src/pjmedia/sdp_neg.c

```diff
@@ -1545,6 +1576,118 @@ static pj_status_t create_answer( pj_pool_t *pool,
     return has_active ? PJ_SUCCESS : status;
 }
 
+/*
+ * Flag:
+ * 1: pt found
+ * 2: codec found
+ * Return (bitwise):
+ * 0: pt and codec not found
+ * 1: only pt found
+ * 2: only codec found
+ * 3: pt and codec found
+ */
+static unsigned pt_exists(const media_pt_map *sess_med_pt_map, unsigned pt, 
+                          const pj_str_t* codec)
+{
+    unsigned i = 0;
+    unsigned cnt = sess_med_pt_map->map_cnt;
+    unsigned max_cnt = PJ_ARRAY_SIZE(sess_med_pt_map->map);
+    unsigned retval = 0;
+    for (; i < cnt && i < max_cnt; ++i) {
+        unsigned tmp_ret = 0;
+
+        if (sess_med_pt_map->map[i].pt == pt) {
+            tmp_ret |= 1;
+        }
+        if (pj_strcmp(&sess_med_pt_map->map[i].codec, codec) == 0) {
+            tmp_ret |= 2;
+        }
+
+        if (tmp_ret == 3)
+            return tmp_ret;
+
+        if (tmp_ret > retval)
+            retval = tmp_ret;
+    }
+    return retval;
+}
+
+#if PJMEDIA_SDP_NEG_MAINTAIN_SESSION_PT
+
+static media_pt_map* get_media_pt_map(pjmedia_sdp_neg* neg, pj_str_t* med_type)
+{
+    unsigned i = 0;
+    unsigned max_cnt = PJ_ARRAY_SIZE(neg->med_pt_map);
+    for (; i < neg->med_pt_map_cnt && i < max_cnt; ++i) {
+        if (pj_strcmp(&neg->med_pt_map[i].med_type, med_type) == 0) {
```

Comment:
the mapping is per stream, not based on media type.

Reply:
considered that too, but what if the media type of a stream is changed?

Reply:
hmm... the RFC doesn't really say, perhaps we can consider resetting the mapping?

But it should be quite a rare scenario, so leaving it as is should also be okay?

A more common scenario is removing and reusing the media slot. So, I guess the mapping needs to be reset upon removal.

Reply:
Media type change is allowed in rfc3264 8.3.3, but yes I believe it should be rare.
But we've also seen media/stream removed which is actually illegal.

Reply:
Consider the scenario:
Offer Media : Audio[1] Video[2] Audio[3]
ReOffer Media : Audio[4] Audio[5]
In this case we need to reset the mapping for Video[2] and Audio[3]. So Audio[4] will use mapping on Audio[1] and Audio[5] will use a new mapping. Or the map will be reset entirely if there's a media removal

Reply:
As usual, let's go back to the RFC:
```
The media type (audio, video, etc.) for a stream MAY be changed.  It
   is RECOMMENDED that the media type be changed (as opposed to adding a
   new stream), when the same logical data is being conveyed, but just
   in a different media format.  This is particularly useful for
   changing between voiceband fax and fax in a single stream, which are
   both separate media types.  To do this, the offerer creates a new
   media description, with a new media type, in place of the description
   in the previous SDP which is to be changed.

   The corresponding media stream in the answer is formulated as
   described in Section 6.  Assuming the stream is acceptable, the
   answerer SHOULD begin sending with the new media type and formats as
   soon as it receives the offer. The offerer MUST be prepared to
   receive media with both the old and new types until the answer is
   received, and media with the new type is received and reaches the top
   of the playout buffer.
```
The RFC only mentions media type change if `same logical data is conveyed`, such as voiceband fax and fax. Thus, drastic media change, such as audio to video or vice versa, although not illegal, should never happen in practice, especially due to requirement such as `The offerer MUST be prepared to receive media with both the old and new types`, which means that it has to maintain both media types during the change.

And even if it does happen, the mapping should stay the same, IMO, since there's chance that it may change back again (med_type1->type2->type1)

In conclusion, IMO, we don't need any special handling for media type change.

Removal of media stream, however, will require the map to be reset, since the RFC says:
```
in the
   case of RTP, the mapping from a particular dynamic payload type
   number to a particular codec within that media stream MUST NOT change
   for the duration of a session
```
If media stream is removed, it no longer exists, so the mapping should also no longer exist.

The PR turns out to be quite a major and complex task, so I will continue the discussion about this in the main PR page.